### PR TITLE
PACT-526 Get record via JMS

### PR DESCRIPTION
**Pull request by felixgb**

### CODE REVIEW

**Comment by tna-erasmos** [comment]
This needs to be formatted, @felixgb .

File: app/uk/gov/nationalarchives/omega/editorial/controllers/EditSetRecordController.scala [hunk 1]
```diff
@@ -49,7 +51,8 @@ class EditSetRecordController @Inject() (
   editSetRecordEdit: editSetRecordEdit,
   editSetRecordEditDiscard: editSetRecordEditDiscard,
   editSetRecordEditSave: editSetRecordEditSave
-) extends MessagesAbstractController(messagesControllerComponents) with I18nSupport with Secured with FormSupport {
+)(implicit val ec: ExecutionContext)
+    extends MessagesAbstractController(messagesControllerComponents) with I18nSupport with Secured with FormSupport {
 
   private val logger: Logger = Logger(this.getClass)
   private lazy val legalStatuses: Seq[LegalStatus] = referenceDataService.getLegalStatuses
```
Comment (tna-erasmos) [inline app/uk/gov/nationalarchives/omega/editorial/controllers/EditSetRecordController.scala:58]: On line 67, we should be using `findRecord`, as well.

File: test/support/TestApiConnector.scala [hunk 2]
```diff
@@ -24,12 +24,17 @@ package support
...
+import uk.gov.nationalarchives.omega.editorial.models.{ EditSet, EditSetRecord }
+import uk.gov.nationalarchives.omega.editorial.{ editSetRecords, editSets }
 
 object TestApiConnector extends ApiConnector(null, null, null) {
 
   override def getEditSet(id: String): IO[EditSet] =
     IO.pure(editSets.editSet1)
 
+  override def getEditSetRecord(editSetOci: String, recordOci: String): IO[Option[EditSetRecord]] =
+    IO.pure {
```
Comment (tna-erasmos) [inline test/support/TestApiConnector.scala:36]: This is at odds with the style above (33); perhaps we should decide on one?

File: app/uk/gov/nationalarchives/omega/editorial/connectors/ApiConnector.scala [hunk 3]
```diff
@@ -85,6 +97,12 @@ class ApiConnector @Inject() (
 
 object ApiConnector {
 
+  sealed abstract class SID(val value: String)
```
Comment (tna-erasmos) [inline app/uk/gov/nationalarchives/omega/editorial/connectors/ApiConnector.scala:100]: Perhaps we should have the `ResponseBuilder` use these?

File: it/ApiConnectorISpec.scala [hunk 4]
```diff
@@ -16,11 +17,32 @@ class ApiConnectorISpec extends BaseISpec {
 
       editSetResponse mustBe editSets.editSet1
     }
+  }
 
-    s"get the edit set 100 times for id: $idOfExistingEditSet" in {
-      val action = (1 to 100).toList.traverse(_ => apiConnector.getEditSet(idOfExistingEditSet))
-
-      action.unsafeRunSync().foreach(_ mustBe editSets.editSet1)
+  "when a request for an editSetRecord is made" must {
```
Comment (tna-erasmos) [inline it/ApiConnectorISpec.scala:22]: Here's a suggested alternative which reduces some duplication:

```
    Set(
      "COAL.2022.V1RJW.P",
      "COAL.2022.V2RJW",
      "COAL.2022.V3RJW.P",
      "COAL.2022.V4RJW.P",
      "COAL.2022.V5RJW.P",
      "COAL.2022.V6RJW.P",
      "COAL.2022.V7RJW.P",
      "COAL.2022.V8RJW.P",
      "COAL.2022.V9RJW.P",
      "COAL.2022.V10RJW.P",
      "COAL.2022.V11RJW.P",
      "COAL.2022.V12RJW.P"
    ).foreach { oci =>
      s"get an edit set record for oci: $oci" in {
        val editSetRecordResponse =
          apiConnector.getEditSetRecord(idOfExistingEditSet, oci).unsafeRunSync()

        editSetRecordResponse mustBe editSetRecords.getEditSetRecordByOCI(oci)
      }
    }
```

**Review by tna-erasmos — CHANGES_REQUESTED** [review]
The only real issue is that there's one other place where this call should be made.

**Comment by tna-erasmos** [comment]
My QA is described [here](https://national-archives.atlassian.net/browse/PACT-526?focusedCommentId=42713).

File: test/support/TestApiConnector.scala (hunk 2 above)
Comment (felixgb) [reply test/support/TestApiConnector.scala:36]: Sure, can do. I think we should stick with the curly brackets, as to me it says "this block has not been executed yet" more.

I'll make that change now.

File: app/uk/gov/nationalarchives/omega/editorial/connectors/ApiConnector.scala (hunk 3 above)
Comment (felixgb) [reply app/uk/gov/nationalarchives/omega/editorial/connectors/ApiConnector.scala:100]: That sounds like a good idea, I'll make that change now.

File: test/support/TestApiConnector.scala (hunk 2 above)
Comment (tna-erasmos) [reply test/support/TestApiConnector.scala:36]: Good point: let's use the curly braces everywhere we use `IO.pure`.

Comment (felixgb) [reply test/support/TestApiConnector.scala:36]: Ah, so scalafmt rewrites curly braces to be parens if it's on a single line. That's a bit agressive...

File: app/uk/gov/nationalarchives/omega/editorial/controllers/EditSetRecordController.scala (hunk 1 above)
Comment (felixgb) [reply app/uk/gov/nationalarchives/omega/editorial/controllers/EditSetRecordController.scala:58]: Done. We were not using `findRecord` in the save or discard actions either.

**Comment by felixgb** [comment]
> This needs to be formatted, @felixgb .

Should I change the CI to reject PRs that are not formatted? That's what I've relied upon in the past.

File: app/uk/gov/nationalarchives/omega/editorial/controllers/EditSetRecordController.scala (hunk 1 above)
Comment (tna-erasmos) [reply app/uk/gov/nationalarchives/omega/editorial/controllers/EditSetRecordController.scala:58]: Oh right.

Comment (tna-erasmos) [reply app/uk/gov/nationalarchives/omega/editorial/controllers/EditSetRecordController.scala:58]: Confirmed.

File: test/support/TestApiConnector.scala (hunk 2 above)
Comment (tna-erasmos) [reply test/support/TestApiConnector.scala:38]: Actually, it's fine as is. Even clearer.
 
Confirmed.

File: it/ApiConnectorISpec.scala (hunk 4 above)
Comment (tna-erasmos) [reply it/ApiConnectorISpec.scala:22]: Yeah, that works, too. I didn't think of using a table like that, but it reduce the duplication.

Small typo on line 47 (`editSetRecordRespose`).

File: app/uk/gov/nationalarchives/omega/editorial/connectors/ApiConnector.scala (hunk 3 above)
Comment (tna-erasmos) [reply app/uk/gov/nationalarchives/omega/editorial/connectors/ApiConnector.scala]: Confirmed.

**Review by tna-erasmos — APPROVED** [review]
Only a typo remaining.

LGTM.

File: app/uk/gov/nationalarchives/omega/editorial/connectors/ApiConnector.scala [hunk 5]
```diff
@@ -85,6 +97,19 @@ class ApiConnector @Inject() (
...
+    def matches(sid: String): Boolean =
+      sid.trim.equalsIgnoreCase(this.value)
+
+  }
+
+  object SID {
+    case object GetEditSet extends SID("OSGEES001")
+    case object GetEditSetRecord extends SID("OSGESR001")
+    case object UpdateEditSetRecord extends SID("OSUESR001")
+  }
```
Comment (rwalpole) [inline app/uk/gov/nationalarchives/omega/editorial/connectors/ApiConnector.scala:111]: You could use an [enumeratum](https://github.com/lloydmeta/enumeratum) for this.

File: app/uk/gov/nationalarchives/omega/editorial/controllers/EditSetRecordController.scala [hunk 6]
```diff
@@ -49,96 +51,98 @@ class EditSetRecordController @Inject() (
...
   private lazy val creators: Seq[Creator] = referenceDataService.getCreators
   private lazy val placesOfDeposit: Seq[PlaceOfDeposit] = referenceDataService.getPlacesOfDeposit
 
-  def viewEditRecordForm(id: String, recordId: String): Action[AnyContent] = Action {
+  def viewEditRecordForm(id: String, recordId: String): Action[AnyContent] = Action.async {
     implicit request: Request[AnyContent] =>
-      withUser { user =>
+      withUserAsync { user =>
         logger.info(s"The edit set id is $id for record id $recordId")
         val editSetName = editSets.getEditSet().name
```
Comment (rwalpole) [inline app/uk/gov/nationalarchives/omega/editorial/controllers/EditSetRecordController.scala:66]: I've raised [another ticket](https://national-archives.atlassian.net/browse/PACT-806) to address this. We don't want to be calling the API to get the whole edit set just to have a name to put into a back link.

**Review by rwalpole — APPROVED** [review]

**Pull request merged**